pywr-core/src/parameters/profiles/weekly.rs

```diff
@@ -0,0 +1,459 @@
+    /// returned.
+    fn next(&self, date: &Date) -> f64 {
+        let current_week_index = self.current_index(date);
+
+        match self {
+            Self::FiftyTwo(values) => {
+                if current_week_index >= 51 {
+                    values[0]
+                } else {
+                    values[current_week_index + 1]
+                }
+            }
+            Self::FiftyThree(values) => {
+                if current_week_index >= 52 {
+                    values[0]
+                } else {
+                    values[current_week_index + 1]
+                }
+            }
+        }
+    }
+
+    /// Get the previous week's value based on the week index of the provided date. If the
+    /// current week index is 0 than the last array value is returned.
+    fn prev(&self, date: &Date) -> f64 {
+        let current_week_index = self.current_index(date);
+
+        match self {
+            Self::FiftyTwo(values) => {
+                if current_week_index == 0 {
+                    values[51]
+                } else {
+                    values[current_week_index - 1]
+                }
+            }
+            Self::FiftyThree(values) => {
+                if current_week_index == 0 {
+                    values[52]
+                } else {
+                    values[current_week_index - 1]
+                }
+            }
+        }
+    }
+
+    /// Find the value corresponding to the given date by linearly interpolating between two
+    /// consecutive week's values.
+    fn interpolate(&self, date: &Date, first_value: f64, last_value: f64) -> f64 {
+        let current_pos = self.current_pos(date);
+        let week_delta = current_pos - current_pos.floor();
+        first_value + (last_value - first_value) * week_delta
+    }
+
+    /// Calculate the value on the given date using the interpolation method option. In a 52-week
+    /// interpolated profile, the upper boundary in the 52nd and 53rd week is the same when
+    /// WeeklyInterpDay is First (i.e. the value on 1st January). When WeeklyInterpDay is Last the
+    /// 1st and last week will share the same lower bound (i.e. the value on the last week).
+    fn value(&self, date: &Date, interp_day: &Option<WeeklyInterpDay>) -> f64 {
+        match interp_day {
+            None => self.current(date),
+            Some(interp_day) => match interp_day {
+                WeeklyInterpDay::First => {
+                    let first_value = self.current(date);
+                    let last_value = self.next(date);
+                    self.interpolate(date, first_value, last_value)
+                }
+                WeeklyInterpDay::Last => {
+                    let first_value = self.prev(date);
+                    let last_value = self.current(date);
+                    self.interpolate(date, first_value, last_value)
+                }
+            },
+        }
+    }
+}
+
+impl From<Vec<f64>> for WeeklyProfileSize {
```

Comment:
In response to my other comment about changing the arguments or return value to `new`, something like this might work:

```rust
#[derive(Error, Debug, PartialEq, Eq)]
pub enum WeeklyProfileError {
    #[error("52 or 53 values must be given for a weekly profile parameter")]
    InvalidLength,
}

impl TryFrom<&[f64]> for WeeklyProfileSize {
    type Error = WeeklyProfileError;

    fn try_from(value: &[f64]) -> Result<Self, Self::Error> {
        match value.len() {
            52 => Ok(WeeklyProfileSize::FiftyTwo(value.try_into().unwrap())),
            53 => Ok(WeeklyProfileSize::FiftyThree(value.try_into().unwrap())),
            _ => Err(WeeklyProfileError::InvalidLength),
        }
    }
}
```

Reply:
Switched to try_from